src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -0,0 +1,434 @@
+            return false;
+        }
+        boost::archive::text_iarchive ar(ifs);
+        ar & storedSeqs_;
+        if (!storedSeqs_.empty())
+        {
+            if (boost::icl::first(storedSeqs_) < firstSeq_ ||
+                boost::icl::last(storedSeqs_) > lastSeq_)
+            {
+                JLOG(j_.error()) <<
+                    "shard " << std::to_string(index_) <<
+                    " invalid control file";
+                return false;
+            }
+            if (boost::icl::length(storedSeqs_) ==
+                (index_ == detail::genesisShardIndex ?
+                    detail::genesisNumLedgers : ledgersPerShard))
+            {
+                JLOG(j_.error()) <<
+                    "shard " << std::to_string(index_) <<
+                    " found control file for complete shard";
+                storedSeqs_.clear();
+                remove(control_);
+                complete_ = true;
+            }
+        }
+    }
+    else
+        complete_ = true;
+    updateFileSize();
+    return true;
+}
+
+bool
+Shard::setStored(std::shared_ptr<Ledger const> const& l)
+{
+    assert(backend_&& !complete_);
+    if (boost::icl::contains(storedSeqs_, l->info().seq))
+    {
+        JLOG(j_.debug()) <<
+            "shard " << std::to_string(index_) <<
+            " ledger seq " << std::to_string(l->info().seq) <<
+            " already stored";
+        return false;
+    }
+    if (boost::icl::length(storedSeqs_) >=
+        (index_ == detail::genesisShardIndex ?
+            detail::genesisNumLedgers : ledgersPerShard) - 1)
+    {
+        if (backend_->fdlimit() != 0)
+        {
+            remove(control_);
+            updateFileSize();
+        }
+        complete_ = true;
+        storedSeqs_.clear();
+
+        JLOG(j_.debug()) <<
+            "shard " << std::to_string(index_) << " complete";
+    }
+    else
+    {
+        storedSeqs_.insert(l->info().seq);
+        lastStored_ = l;
+        if (backend_->fdlimit() != 0 && !saveControl())
+            return false;
+    }
+
+    JLOG(j_.debug()) <<
+        "shard " << std::to_string(index_) <<
+        " ledger seq " << std::to_string(l->info().seq) <<
+        " stored";
+
+    return true;
+}
+
+boost::optional<std::uint32_t>
+Shard::prepare()
+{
+    if (storedSeqs_.empty())
+         return lastSeq_;
+    return prevMissing(storedSeqs_, 1 + lastSeq_, firstSeq_);
+}
+
+bool
+Shard::hasLedger(std::uint32_t seq) const
+{
+    if (seq < firstSeq_ || seq > lastSeq_)
+        return false;
+    if (complete_)
+        return true;
+    return boost::icl::contains(storedSeqs_, seq);
+}
+
+void
+Shard::validate(Application& app)
+{
+    uint256 hash;
+    std::uint32_t seq;
+    std::shared_ptr<Ledger> l;
+    // Find the hash of the last ledger in this shard
+    {
+        std::tie(l, seq, hash) = loadLedgerHelper(
+            "WHERE LedgerSeq >= " + std::to_string(lastSeq_) +
+            " order by LedgerSeq desc limit 1", app);
```

Comment:
I made that exact mistake, I sorted with ASC instead of DESC and the problem is we are less likely to have the older ledger data in the sqllite DB. That was the reason we were occasionally getting the no data available errors when trying to validate. Using the latest data solves that issue.

Reply:
We also want to be able to run this code as soon as we have a valid ledger. In that case, we haven't validated any older data yet and it can be incomplete or erroneous. Our latest ledger is sure to be valid/complete if we think we're in synch or have ever been in synch in this run.